encoding/template/run_length_decode.go

```diff
@@ -0,0 +1,83 @@
+package template
+
+import (
+	"encoding/binary"
+	"errors"
+
+	"github.com/xichen2020/eventdb/x/io"
+)
+
+var (
+	errInvalidNumberOfRepetitions = errors.New("invalid # of repetitions < 1")
+)
+
+// ReadValueFn reads a GenericValue from an `io.Reader`.
+type ReadValueFn func(reader io.Reader) (GenericValue, error)
+
+// runLengthDecodeValue run length decodes a stream of GenericValues.
+func runLengthDecodeValue(
+	reader io.Reader,
+	ReadValue ReadValueFn,
+) *RunLengthValueIterator {
+	return newRunLengthValueIterator(reader, ReadValue)
+}
+
+// RunLengthValueIterator iterates over a run length encoded stream of GenericValue data.
+type RunLengthValueIterator struct {
+	reader      io.Reader
+	readValue   ReadValueFn
+	curr        GenericValue
+	repetitions int64
```

Comment:
`int` would probably be more idiomatic.

Reply:
I made this `int64` so I could read directly into it using `binary.ReadVarint`.

Reply:
Hm sure.